✨ `BaseRestartWorkChain`: add `pause_on_max_iterations` input

mbercx opened the pull request:
@cpignedoli this is a rough draft PR for you to test! I have to sign off for the day, but already wanted to get your feedback.

To be clear: this is not ready for proper review. ^^ It's based on/blocked by https://github.com/aiidateam/aiida-core/pull/7116, `pre-commit` was ignored, no tests, no documentation changes. Just to see if it already works as expected and to get feedback on one question: 

1. if the process is paused because the max iterations of a handler was hit, which counters do we reset? Right now, only the corresponding handler counter is reset, or the global one in case that was hit. Should all counters be reset? Or should we reset the global counter as well in case we hit the max iterations of a single error handler?

Some code that may help in testing:

```python
from aiida import orm, engine, load_profile
from aiida_quantumespresso.workflows.pw.base import PwBaseWorkChain
from ase.build import bulk

load_profile('dev')

code = orm.load_code("pw@localhost")

structure = orm.StructureData(ase=bulk('Si'))

builder = PwBaseWorkChain.get_builder_from_protocol(
    code, structure,
    protocol='precise'
)
# builder.max_iterations = 1
builder.handler_overrides = {
    'handle_out_of_walltime': {
        'max_iterations': 1
    }
}
builder.pause_on_max_iterations = True
builder.pw.parameters['CONTROL']['max_seconds'] = 3

engine.submit(builder)
```

cpignedoli commented:
Hi @mbercx I tested it directly in the QE App aiida 2.7.1 aiida_quantumespresso 4.13
```
            relax_builder.base.handler_overrides = {
                'handle_out_of_walltime': {
                    'max_iterations': 1
                }
            }
            relax_builder.base.pause_on_max_iterations = True
```
For me it works, just a minor bug (see image below):
in re report it writes "5" as max iterations, probably because it takes it from self.inputs, and you redefine a different variable
```
max_handler_iterations = max_iterations_override if max_iterations_override else handler.max_iterations
```
<img width="756" height="179" alt="image" src="https://github.com/user-attachments/assets/eee0c60a-b675-4235-9e23-fa72602ef58d" />

mbercx commented:
Thanks for testing @cpignedoli! Good catch on the report. I fixed this, but realised a critical issue in my original implementation: 

Returning an exit code in the `should_run_process` method - which is used in a `while_` loop in the outline - is interpreted as `True`. This breaks the logic: another process will be run even though an exit code should be returned, and the workflow keeps on restarting even though the max iterations is reached. I'm not sure if this _should_ be the case, but that is a discussion on `plumpy` that is beyond the scope of this PR.

To fix this, I've moved all the logic regarding `max_iterations` to where I think it should be: the `inspect_process` step. Some other improvements:

1. The report now no longer claims to be restarting, then aborts because the max iterations has been reached. On the current `main` branch, you get (trimmed the prefix):

   ```
   ...]: PwCalculation<1560> failed but a handler dealt with the problem, restarting
   ...]: reached the maximum number of iterations 1: last ran PwCalculation<1560>
   ```

   Now you get:

   ```
   ...]: Reached the maximum number of global iterations (1).
   ...]: Aborting! Last ran: PwCalculation<1577>
   ```

2. When pausing, the logic of resetting the counters is improved. Basically, the code checks each of the max iterations: global, and for each of the handlers. If `max_iterations` is hit and `pause_on_max_iterations` is `False`, the work chain aborts. Else the corresponding counter is already reset and the `pause_process` variable is set to `True`. After checking all max iterations the process is then paused. This gives a clear report:

   ```
   ...]: Reached the maximum number of global iterations (1).
   ...]: Reached the maximum number of iterations (1) for handler `handle_out_of_walltime`.
   ...]: PwCalculation<1614> failed but a handler dealt with the problem, restarting
   ...]: Resetting the iteration counter(s) and pausing for inspection. You can resume execution using `verdi process play 1609`, or kill the work chain using `verdi process kill 1609`.
   ```

   And because _both_ the handler and global counters are reset in this case, you don't get the issue that the process will always pause again the next iteration because one of the two is still hitting `max_iterations`. However, other counters are unaffected. I think this is the correct behavior.

Tests and documentation have been added, so I think this one is now ready for proper review!

mbercx commented on src/aiida/engine/processes/workchains/restart.py:
```diff
@@ -375,9 +417,20 @@ def inspect_process(self) -> Optional['ExitCode']:
+                )
+                self.pause(f"Paused for user inspection, see: 'verdi process report {self.node.pk}'")
 
-            return last_report.exit_code
```
Here, the `BaseRestartWorkChain` used the fact that returning `ExitCode(0)` in a step of the outline doesn't really do anything as far as I can tell: it simply starts executing the next step. Might be there is not even any difference between returning that and `None`.

I think it's better to be explicit and return the exit code in case its exit status is nonzero, and return `None` otherwise.

mbercx commented on src/aiida/engine/processes/workchains/restart.py:
```diff
@@ -210,15 +217,14 @@ def setup(self) -> None:
+        This is the case as long as the last process has not finished successfully or a handler has been triggered.
         """
-        max_iterations = self.inputs.max_iterations.value
-        return not self.ctx.is_finished and self.ctx.iteration < max_iterations
```
Whether or not the global `max_iterations` has been reached is now checked in the `inspect_process` step.

cpignedoli commented:
Hi @mbercx, testing right now with QE app, same config as above.
I am using SiO2, Fast, 30 seconds, allowing for 3 iterations.

# Restart three times and exit OK

<img width="1173" height="586" alt="image" src="https://github.com/user-attachments/assets/81ed6676-0fcf-4719-8b14-1c689ee35deb" />

# Restart three times, pause, restart three times OK

<img width="1162" height="655" alt="image" src="https://github.com/user-attachments/assets/b246942a-0ed8-4c54-b3d5-91577f9958e5" />

<img width="722" height="603" alt="image" src="https://github.com/user-attachments/assets/ee2083ca-0024-437e-abe0-b249cfa1bf56" />

# Restart + unhandled 
Not sure if the unhandled_error logic is supposed to work here, but it does not if I try the same way I was doing for that PR:
<img width="479" height="118" alt="image" src="https://github.com/user-attachments/assets/a7045ea9-7af0-4fad-89e5-17ddce6fce1e" />
I let it run out of walltime, then I manually killed the pw.x process, and the unhandled error causes aborting:
<img width="427" height="82" alt="image" src="https://github.com/user-attachments/assets/2714120a-384b-48e1-a103-a27ffcf1a949" />

GeigerJ2 commented on src/aiida/engine/processes/workchains/restart.py:
```diff
@@ -364,8 +370,51 @@ def inspect_process(self) -> Optional['ExitCode']:
         self.ctx.unhandled_failure = False
 
-        # If at least one handler returned a report, the action depends on its exit code and that of the process itself
         if last_report:
```
A bit of a scope creep, but maybe we can invert this condition to use an early return to reduce the nesting level and cognitive complexity of this method?

GeigerJ2 commented on src/aiida/engine/processes/workchains/restart.py:
```diff
@@ -364,8 +370,51 @@ def inspect_process(self) -> Optional['ExitCode']:
+                max_iterations_override = self.ctx.handler_overrides.get(handler.__name__, {}).get(
+                    'max_iterations', None
+                )
+                max_handler_iterations = max_iterations_override if max_iterations_override else handler.max_iterations  # type: ignore[attr-defined]
```
```suggestion
                max_handler_iterations = max_iterations_override or handler.max_iterations  # type: ignore[attr-defined]
```

GeigerJ2 commented on src/aiida/engine/processes/workchains/restart.py:
Hm, this method is now ~200 lines and quite difficult to follow the high-level logic (at least with a modern-day attention span of 5s :D). Maybe some parts of the logic can be extracted to helper functions so that the high-level flow is easier to follow?

GeigerJ2 commented on src/aiida/engine/processes/workchains/restart.py:
```suggestion
    values should be a dictionary that can specify the keys ``enabled`` and ``priority``, and ``max_iterations``.
```

GeigerJ2 commented on src/aiida/engine/processes/workchains/utils.py:
```suggestion
        which has status `0` meaning that the work chain will restart the process
        (unless max iterations is reached).
```

GeigerJ2 reviewed (approved):
As always, very clean, thanks @mbercx! Just a few doc(string) updates to reflect the updated logic. Already approving :)

EDIT: Not sure why my suggestions don't appear properly on the lines that I propose to edit (might be an issue with the browser...), but you get the gist.

GeigerJ2 reviewed (approved):
Let's get this baby in, thanks @mbercx!

mbercx committed:
✨ `BaseRestartWorkChain`: add `pause_on_max_iterations` input

The `BaseRestartWorkChain` limits the maximum number of iterations via the
`max_iterations` input, to avoid the workflow from indefinitely trying to recover a
failing calculation via ineffective handling and using resources unnecessarily. When
reaching the maximum number of iterations, the workflow will currently simply abort
with the `ERROR_MAXIMUM_ITERATIONS_EXCEEDED` exit code.

In some cases however, a user may want to inspect the workflow and determine if more
restarts are desirable. A common example here is when running a long calculation that
hits the walltime multiple times. In this case, it's possible that a few more iterations
would do the job. Simply increasing the `max_iterations` for all runs might also lead to
other runs using too many resources.

To allow the user to inspect the workflow and decide how to proceed, we add the
`pause_on_max_iterations` boolean input. If set to `True`, the workflow will pause when
the maximum iterations is reached, and the iteration counter is reset. The user can then
decide to continue (play) the workflow, or abort using the kill command.

The logic on dealing with the max iterations is also isolated on the `inspect_process`
step. This is because in case the user decides to pause the process, we still need to
be able to restart and hence can't add the logic to the `results` step. The
`should_run_process` step is also not an option, since it is used in a `_while` loop and
hence exit codes returned here are simply interpreted as `True`. By moving all the logic
to the `inspect_process` step, it's also more localized and easier to understand all
possibilities.

GeigerJ2 reviewed (approved):
LGTM, thanks, @mbercx! Just one small fix I'd still recommend:

The `self.ctx.iteration = 0` reset (line 392) happens before we check if the handler indicates an unrecoverable failure. If `pause_on_max_iterations=True` but the handler returns `exit_code.status != 0`, we abort at line 407 with the counter already reset to 0. This might eb confusing for debugging failed workchains.

Instead, one can move `self.ctx.iteration = 0` into the `if pause_process:` block at line 409, so the counter is only reset when we actually pause.

Otherwise good to merge!

mbercx committed:
Apply review 1
